### ISSUE TITLE
sentence probability (for sentence level training)

### ISSUE THREAD

**Issue by borgr** [OpenNMT/OpenNMT-py#904]
I am trying to create a loss function based on the the probability of a whole sentence (such example would be MRT). I know it is not supported in the loss functions, so I want to train a copy attention model normally (word cross entropy) and then load the last model and make my own addition.
I am trying to understand how can I get a probability of a given sentence.

In a pseudo/high level code I would:
load model
sum = 0
for i in range (0.. len(target_sentence)):
 sum += the probability of creating target_sentence[i] given predicted target_sentence[:i]

add some losses and apply gradients

**Comment by chrishokamp** [comment]
Hi, I think you should be able to get the (log) probability `p(y|x)` by calling `model.generator(outputs)` on the outputs of the model (like after these lines): https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/trainer.py#L281-L282). So, for example, if you paste the following code at [line 283](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/trainer.py#L283), it should print the log-probability of each target sequence (Warning: written quickly, could be wrong, syntax could definitely be simplified):

```

import torch

token_log_probs = self.model.generator(outputs)
# time, batch, target_vocab_size
t, b, v = token_log_probs.shape

# get the probabilities of each target token `p(y_t | x; y_0...y_t-1)`
tgt_log_probs = token_log_probs.view(t*b, -1)[torch.arange(t*b).long(), batch.tgt[1:].view(-1)].view(t, b)
# zero out padding idx
padding_idxs = batch.tgt[1:].eq(batch.dataset.fields['tgt'].vocab.stoi[inputters.PAD_WORD])
tgt_log_probs[padding_idxs] = 0.

# sum along the time dimension to get `log(p(y|x))`
seq_probs = tgt_log_probs.sum(dim=0)
print(seq_probs)

# compute your loss function, run backward, step optimizer, etc... 
```

**Comment by xptree** [comment]
> I am trying to create a loss function based on the the probability of a whole sentence (such example would be MRT). I know it is not supported in the loss functions, so I want to train a copy attention model normally (word cross entropy) and then load the last model and make my own addition.
> I am trying to understand how can I get a probability of a given sentence.
> 
> In a pseudo/high level code I would:
> load model
> sum = 0
> for i in range (0.. len(target_sentence)):
> sum += the probability of creating target_sentence[i] given predicted target_sentence[:i]
> 
> add some losses and apply gradients

@borgr Hi, I wonder if you have successfully implemented loss like MRT with opennmt. If the answer is yes, is the code open-sourced?

Thanks!

**Comment by borgr** [comment]
Used NEMATUS, there is also beyond-bleu if you prefer pytorch
